Error message:
```
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.10/dist-packages/paddle/io/dataloader/dataloader_iter.py", line 604, in _thread_loop
    batch = self._get_data()
  File "/usr/local/lib/python3.10/dist-packages/paddle/io/dataloader/dataloader_iter.py", line 723, in _get_data
    raise e
  File "/usr/local/lib/python3.10/dist-packages/paddle/io/dataloader/dataloader_iter.py", line 693, in _get_data
    data = self._data_queue.get(timeout=self._timeout)
  File "/usr/lib/python3.10/multiprocessing/queues.py", line 122, in get
    return _ForkingPickler.loads(res)
  File "/usr/local/lib/python3.10/dist-packages/paddle/incubate/multiprocessing/reductions.py", line 87, in _rebuild_tensor
    tensor.value().get_tensor()._share_data_with(lodtensor)
  File "/usr/local/lib/python3.10/dist-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/usr/local/lib/python3.10/dist-packages/paddle/fluid/wrapped_decorator.py", line 25, in __impl__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/paddle/fluid/framework.py", line 445, in __impl__
    assert in_dygraph_mode(), (
AssertionError: We only support 'value()' in dynamic graph mode, please call 'paddle.disable_static()' to enter dynamic graph mode.
LAUNCH INFO 2023-12-10 03:02:36,536 Exit code 1
[2023-12-10 03:02:36,536] [    INFO] controller.py:149 - Exit code 1
```

--- a/test/legacy_test/test_multiprocess_dataloader_dynamic.py
+++ b/test/legacy_test/test_multiprocess_dataloader_dynamic.py
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import logging
 import sys
 import time
 import unittest
@@ -32,6 +33,8 @@ import paddle
 from paddle import base
 from paddle.io import DataLoader
 from paddle.nn import Linear
+
+logging.basicConfig(format="%(asctime)s - %(levelname)s - %(message)s")
 
 
 class SimpleFCNet(paddle.nn.Layer):
@@ -74,8 +77,24 @@ class SimpleFCNet(paddle.nn.Layer):
         return out
 
 
+def collate_batch(batch_list):
+    batch_size = len(batch_list)
+    image = np.stack([item[0] for item in batch_list], axis=0).astype('float32')
+    image = paddle.to_tensor(image).reshape([batch_size, -1])
+    label = np.stack([item[1] for item in batch_list], axis=0).astype('int64')
+    label = paddle.to_tensor(label).reshape([batch_size, -1])
+    return image, label
+
+
 class TestDygraphDataLoader(unittest.TestCase):
-    def run_main(self, num_workers, places, persistent_workers):
+    def run_main(
+        self,
+        num_workers,
+        places,
+        persistent_workers,
+        collate_fn,
+        use_shared_memory,
+    ):
         base.default_startup_program().random_seed = 1
         base.default_main_program().random_seed = 1
         with base.dygraph.guard(places[0]):
@@ -89,6 +108,8 @@ class TestDygraphDataLoader(unittest.TestCase):
                 batch_size=BATCH_SIZE,
                 drop_last=True,
                 persistent_workers=persistent_workers,
+                collate_fn=collate_fn,
+                use_shared_memory=use_shared_memory,
             )
             assert len(dataloader) == int(SAMPLE_NUM / BATCH_SIZE)
 
@@ -117,36 +138,44 @@ class TestDygraphDataLoader(unittest.TestCase):
             "step": step_list,
             "loss": np.array(loss_list),
         }
-        print("time cost", ret['time'], 'step_list', ret['step'])
+        logging.info(f"time cost {ret['time']} step_list {ret['step']}")
         return ret
 
     def test_main(self):
         for p in prepare_places():
             for persistent_workers in [False, True]:
-                results = []
-                for num_workers in [0, 2]:
-                    print(
-                        self.__class__.__name__,
-                        p,
-                        num_workers,
-                        persistent_workers,
-                    )
-                    sys.stdout.flush()
-                    ret = self.run_main(
-                        num_workers=num_workers,
-                        places=p,
-                        persistent_workers=persistent_workers,
-                    )
-                    results.append(ret)
-                diff = np.max(
-                    np.abs(results[0]['loss'] - results[1]['loss'])
-                    / np.abs(results[0]['loss'])
-                )
-                self.assertLess(diff, 1e-2)
+                for collate_fn in [None, collate_batch]:
+                    for use_shared_memory in [False, True]:
+                        results = []
+                        for num_workers in [0, 2]:
+                            logging.info(
+                                f"{self.__class__.__name__} {p} {num_workers} {persistent_workers} {collate_fn} {use_shared_memory}"
+                            )
+                            sys.stdout.flush()
+                            ret = self.run_main(
+                                num_workers=num_workers,
+                                places=p,
+                                persistent_workers=persistent_workers,
+                                collate_fn=collate_fn,
+                                use_shared_memory=use_shared_memory,
+                            )
+                            results.append(ret)
+                        diff = np.max(
+                            np.abs(results[0]['loss'] - results[1]['loss'])
+                            / np.abs(results[0]['loss'])
+                        )
+                        self.assertLess(diff, 1e-2)
 
 
 class TestDygraphDataLoaderWithBatchedDataset(TestDygraphDataLoader):
-    def run_main(self, num_workers, places, persistent_workers):
+    def run_main(
+        self,
+        num_workers,
+        places,
+        persistent_workers,
+        collate_fn,
+        use_shared_memory,
+    ):
         base.default_startup_program().random_seed = 1
         base.default_main_program().random_seed = 1
         with base.dygraph.guard(places[0]):
@@ -160,6 +189,8 @@ class TestDygraphDataLoaderWithBatchedDataset(TestDygraphDataLoader):
                 batch_size=None,
                 drop_last=True,
                 persistent_workers=persistent_workers,
+                collate_fn=None,
+                use_shared_memory=use_shared_memory,
             )
             assert len(dataloader) == int(SAMPLE_NUM / BATCH_SIZE)
 
@@ -188,7 +219,7 @@ class TestDygraphDataLoaderWithBatchedDataset(TestDygraphDataLoader):
             "step": step_list,
             "loss": np.array(loss_list),
         }
-        print("time cost", ret['time'], 'step_list', ret['step'])
+        logging.info(f"time cost {ret['time']} step_list {ret['step']}")
         return ret
 
 

